holoviews/operation/element.py

```diff
@@ -694,17 +699,24 @@ def _process(self, p, element, ranges={}):
         permuted_dims = [(d1, d2) for d1 in dims
                          for d2 in dims[::-1]]
 
+        # Handle histograms as a special case
+        if p.diagonal_type is Histogram:
+            p.diagonal_type = None
+            p.diagonal_operation = histogram.instance(adjoin=False)
+
         data = {}
         for d1, d2 in permuted_dims:
             if d1 == d2:
-                if p.diagonal_type is Histogram:
-                    bin_range = ranges.get(d1.name, element.range(d1))
-                    el = element.hist(dimension=d1.name,
-                                      bin_range=bin_range,
-                                      adjoin=False)(norm=dict(axiswise=True, framewise=True))
-                else:
+                if p.diagonal_type is not None:
                     values = element.dimension_values(d1)
                     el = p.diagonal_type(values, vdims=[d1])
+                elif isinstance(p.diagonal_operation, histogram):
+                    bin_range = ranges.get(d1.name, element.range(d1))
+                    el = p.diagonal_operation(element,
+                                              dimension=d1.name,
+                                              bin_range=bin_range)
```

Comment:
I'm afraid this won't behave very well unless axiswise normalization is enabled here.